### DISCUSSION TITLE
Channel Factory Implementation for CoreWCF

### DISCUSSION THREAD

**Discussion by NiikhilMohanG** [CoreWCF/CoreWCF#1161] [Q&A] [answered]
Hi,
I utilised the CoreWCF namespace's ServiceContract and OperationContract in my code, but when I attempted to instantiate ChannelFactory, I saw an error stating that they were not inherited from ServiceContract. After some research, I discovered the solution: in order for the ChannelFactory implementation to function, we must use the ServiceContract and OperationContract from System.ServiceModel. But in my situation, I must utilise ServiceContracts that are supported by CoreWCF. 
Is there a CoreWCF implementation available which is equivalent to the ChannelFactory?

**Comment by mconnew** [comment]
If you aren't sharing code between the client and your service, there are contract attributes in the CoreWCF namespace. If you are going to share code, you can add a package reference to the System.ServiceModel.Primitives package and use the attributes in the System.ServiceModel namespace and CoreWCF will recognize them. The attributes in the CoreWCF namespace are so you can avoid unnecessarily adding a reference to the client packages in your service app.

**Reply by NiikhilMohanG** [reply]
Thanks. One more query, how can I go about writing some hosted tests against my service? I discovered some client tests utilising HttpClient in the CoreWCF source code. Can we test the services in any other method other that?

In my scenario, I have added project reference of the service to the test project. Then with the help of host builder I started the service. Is there any way for me to test the service other than using HttpClient calls? Since the ChannelFactory implementation wont work as I am using the contract attributes from CoreWCF namespace.

**Reply by mconnew** [reply, marked as answer]
Create a second version of the interface with the System.ServiceModel attributes. All the attributes do is provide WCF with some of the information that's needed to know what XML is expected in the SOAP message. As long as the attributes have the same information, then the client can use what's technically a different interface type, but is equivalent. Just copy the service contract interface into a different .NET namespace for the client and add the System.ServiceModel namespaced attributes with the same property values. There's some default values automatically generated but as long as you keep the actual name of the interface the same (but different namespace), then I believe the generated values will end up the same. If you are specifying an XML namespace via the `ServiceContractAttribute.Namespace` property, then this needs to remain the same.
